File: camerax_android/lib/src/ml/barcode/barcode_scanner_impl.dart
import 'package:camerax_android/src/api.dart';
import 'package:camerax_android/src/ml.dart';
import 'package:camerax_platform_interface/camerax_platform_interface.dart';

final class BarcodeScannerImpl extends DetectorImpl<List<Barcode>>
    implements BarcodeScanner {
  @override
  final BarcodeScannerProxyApi api;

  BarcodeScannerImpl.internal(this.api);

  @override
  Future<List<Barcode>> process(InputImage image) =>
      api.process(image.api).then((e) => e.map((e1) => e1.impl).toList());
}

final class BarcodeScannerChannelImpl extends BarcodeScannerChannel {
  @override
  BarcodeScanner create() {
    final api = BarcodeScannerProxyApi();
    return BarcodeScannerImpl.internal(api);
  }

  @override
  BarcodeScanner createWithOptions(BarcodeScannerOptions options) {
    final api = BarcodeScannerProxyApi.options(options: options.api);
    return BarcodeScannerImpl.internal(api);
  }
}

extension BarcodeScannerX on BarcodeScanner {
  BarcodeScannerProxyApi get api {
    final impl = this;
    if (impl is! BarcodeScannerImpl) throw TypeError();
    return impl.api;
  }
}

File: camerax_android/android/src/main/kotlin/dev/hebei/camerax_android/ml/barcode/BarcodeScannerImpl.kt
package dev.hebei.camerax_android.ml.barcode

import com.google.android.odml.image.MlImage
import com.google.mlkit.vision.barcode.BarcodeScannerOptions
import com.google.mlkit.vision.barcode.BarcodeScanning
import com.google.mlkit.vision.barcode.common.Barcode
import com.google.mlkit.vision.common.InputImage
import dev.hebei.camerax_android.CameraXImpl
import dev.hebei.camerax_android.PigeonApiBarcodeScannerApi
import dev.hebei.camerax_android.ml.Detector

class BarcodeScannerImpl(impl: CameraXImpl) : PigeonApiBarcodeScannerApi(impl) {
    override fun pigeon_defaultConstructor(options: BarcodeScannerOptions?): BarcodeScanner {
        val instance = if (options == null) BarcodeScanning.getClient()
        else BarcodeScanning.getClient(options)
        return BarcodeScanner(instance)
    }

    override fun process0(pigeon_instance: BarcodeScanner, image: MlImage, callback: (Result<List<Barcode>>) -> Unit) {
        pigeon_instance.instance.process(image).addOnSuccessListener { value -> callback(Result.success(value)) }
            .addOnFailureListener { exception -> callback(Result.failure(exception)) }
    }

    override fun process1(
        pigeon_instance: BarcodeScanner, image: InputImage, callback: (Result<List<Barcode>>) -> Unit
    ) {
        pigeon_instance.instance.process(image).addOnSuccessListener { value -> callback(Result.success(value)) }
            .addOnFailureListener { exception -> callback(Result.failure(exception)) }
    }
}

class BarcodeScanner internal constructor(override val instance: com.google.mlkit.vision.barcode.BarcodeScanner) :
    Detector(instance)